docs/site/Express-middleware.md

```diff
@@ -0,0 +1,229 @@
+  // ...
+}
+```
+
+### Define a provider class to allow dependency injection of middleware configuration
+
+```ts
+/**
+ * Define a provider class that wraps the middleware as an interceptor
+ * @param middleware - Middleware module name or factory function
+ * @param className - Class name for the generated provider class
+ *
+ * @typeParam CFG - Configuration type
+ * @typeParam CTX - Context type
+ */
+export function defineInterceptorProvider<
+  CFG,
+  CTX extends Context = InvocationContext
+>(
+  middleware: string | MiddlewareFactory<CFG>,
+  className?: string,
+): Constructor<Provider<GenericInterceptor<CTX>>> {
+  // ...
+}
+```
+
+Alternatively, we can create a subclass of `MiddlewareInterceptorProvider`.
+
+### Use middleware as invocation interceptors
+
+With the ability to wrap Express middleware as LoopBack 4 interceptors, we can
+use the same programming model to register middleware as global interceptors or
+local interceptors denoted by `@intercept` decorators at class and method
+levels.
+
+### Introduce middleware to REST sequence of actions
+
+There are a few steps involved in the default sequence of actions. See
+`Sequence.md` for more details.
+
+It's often desirable to reuse Express middleware in the sequence to handle API
+requests/responses without reinventing the world. We now add an `Middleware`
+action as the first step in the default sequence. The action itself is an
+interceptor chain of `RequestContext`. It uses the powerful
+[extension point/extension pattern](Extension-point-and-extensions.md) to
+contribute and discover registered Express middleware.
+
+#### Register an Express middleware for the sequence
+
+````ts
+/**
+ * Bind a middleware interceptor to this server context
+ *
+ * @example
+ * ```ts
+ * server.middleware('express-middleware-name', {});
+ * ```
+ * @param middleware - Middleware module name or factory function
+ * @param middlewareConfig - Middleware config
+ * @param options - Options for registration
+ *
+ * @typeParam CFG - Configuration type
+ */
+ middleware<CFG>(
+   middleware: string | MiddlewareFactory<CFG>,
+   middlewareConfig?: CFG,
+   options: MiddlewareActionBindingOptions = {},
+ ): Binding<RequestInterceptor> {
+   // ...
+ }
+````
+
+```ts
+const binding = app.middleware(spy, spyConfig, {
+  injectConfiguration: true,
+  key: 'interceptors.middleware.spy',
+});
+```
+
+#### Default sequence
+
+We add a new `middleware` property to `DefaultSequence` to allow injection of
+the middleware action while keeping backward compatibility.
+
+```ts
+export class DefaultSequence implements SequenceHandler {
+  /**
+   * Optional middleware chain
+   * Invokes registered middleware (injected via SequenceActions.MIDDLEWARE).
+   */
+  @inject(SequenceActions.MIDDLEWARE, {optional: true})
+  protected middleware?: Middleware;
+
+  // ...
+  async handle(context: RequestContext): Promise<void> {
+    try {
+      const {request, response} = context;
+      if (this.middleware != null) {
```

Comment:
Instead of using an `if` block here, can you set `middleware` to a no-op function in the constructor when no value is provided by dependency injection?